Create tests

shntnu opened the pull request:
@johnarevalo I am trying to move closer to being able to test this so that we can make changes more confidently

- Added a test
- Copied files from `input`, `metadata`, etc. into the `tests/fixtures` folder to keep them fixed
- Copied 2 CRISPR plates (and just a few feature columns) to test the CRISPR pipeline, which likely covers most of the functions

---

Keep an eye out for this:

To run the tests, first set your PYTHONPATH to include the repository root:

```bash
export PYTHONPATH=$(pwd)
pytest
```

---

```
pytest --cov=jump_profiling_recipe --cov-report=term-missing
=================================================================================== test session starts ===================================================================================
platform darwin -- Python 3.11.11, pytest-8.3.4, pluggy-1.5.0
rootdir: /Users/shsingh/Documents/GitHub/jump-profiling-recipe
configfile: pytest.ini
plugins: cov-6.0.0
collected 1 item                                                                                                                                                                          

tests/test_pipeline_integration.py .                                                                                                                                                [100%]

--------- coverage: platform darwin, python 3.11.11-final-0 ----------
Name                                                                        Stmts   Miss Branch BrPart   Cover   Missing
------------------------------------------------------------------------------------------------------------------------
src/jump_profiling_recipe/__init__.py                                           0      0      0      0 100.00%
src/jump_profiling_recipe/correct/__init__.py                                   4      0      0      0 100.00%
src/jump_profiling_recipe/correct/corrections.py                              156     12     28      9  86.41%   49, 65, 96, 152-170, 218->255, 243, 249->255, 255->265, 259->265, 265->271, 500
src/jump_profiling_recipe/correct/format_check.py                              46      1     12      2  94.83%   58->65, 128
src/jump_profiling_recipe/correct/harmony.py                                   17      1      4      1  90.48%   51
src/jump_profiling_recipe/preprocessing/__init__.py                             9      0      0      0 100.00%
src/jump_profiling_recipe/preprocessing/clean.py                               26      2      2      1  89.29%   37-38
src/jump_profiling_recipe/preprocessing/feature_selection.py                   32      0      2      1  97.06%   58->64
src/jump_profiling_recipe/preprocessing/io.py                                 126     11     30      8  86.54%   50, 52, 54, 94-96, 104, 180, 198-199, 255
src/jump_profiling_recipe/preprocessing/metadata.py                            72     14     12      4  78.57%   130, 146, 171, 206-215, 263-264, 314-315, 319->326
src/jump_profiling_recipe/preprocessing/metrics.py                             56      0      4      0 100.00%
src/jump_profiling_recipe/preprocessing/normalize.py                           21      0      0      0 100.00%
src/jump_profiling_recipe/preprocessing/sphering.py                            32      0      2      0 100.00%
src/jump_profiling_recipe/preprocessing/stats.py                               91     12      6      2  83.51%   50-55, 137-139, 165, 169-170
src/jump_profiling_recipe/preprocessing/transform.py                           21     15      2      0  26.09%   40-55, 73-79
src/jump_profiling_recipe/preprocessing/utils.py                                4      1      2      1  66.67%   21
src/jump_profiling_recipe/pycytominer/operations/__init__.py                    4      0      0      0 100.00%
src/jump_profiling_recipe/pycytominer/operations/correlation_threshold.py      33      9      8      3  65.85%   34, 82, 104-108, 127-135
src/jump_profiling_recipe/pycytominer/operations/transform.py                  52     18     26      9  55.13%   53, 60, 84, 90-94, 105, 117-128, 135->143, 148-153, 175-176
src/jump_profiling_recipe/pycytominer/operations/variance_threshold.py         30      6      4      2  76.47%   46, 94-95, 98-99, 104
------------------------------------------------------------------------------------------------------------------------
TOTAL                                                                         832    102    144     43  83.09%

```

shntnu commented:
I realized we should pull the metadata directly from the library rather than creating a separate copy in the test fixtures

shntnu commented:
PS - @johnarevalo  you might like to run `pytest --cov=jump_profiling_recipe --cov-report=html` to get additional insight when reviewing this PR

johnarevalo commented:
Thanks for setting all this. I tried to run this in my pc, but mamba failed to reproduce the environment. Instead, I moved it to `uv` as that's what we are now using inside the lab. I also updated snakemake to the most recent stable version.

There's some issue with some numbers in the output of harmony-pca. seems some of the PCs are flipped `- (x)`,thus the `pd.assert_equal` statement fails.

shntnu commented:
> Thanks for setting all this. I tried to run this in my pc, but mamba failed to reproduce the environment. Instead, I moved it to `uv` as that's what we are now using inside the lab. I also updated snakemake to the most recent stable version.

`datrie` installation failed at first:

```
uv sync

Using CPython 3.13.0
Creating virtual environment at: .venv
Resolved 115 packages in 1ms
error: Failed to prepare distributions
  Caused by: Failed to download and build `datrie==0.8.2`
  Caused by: Build backend failed to build wheel through `build_wheel` (exit status: 1)

[stdout]
running bdist_wheel
running build
running build_clib
building 'datrie' library
creating build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/alpha-map.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/alpha-map.o
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/darray.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/darray.o
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/dstring.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/dstring.o
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/fileutils.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/fileutils.o
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/tail.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/tail.o
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/trie-string.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/trie-string.o
clang -fno-strict-overflow -Wsign-compare -Wunreachable-code -DNDEBUG -g -O3 -Wall -arch arm64 -mmacosx-version-min=11.0 -Wno-nullability-completeness -Wno-expansion-to-defined -Wno-undef-prefix -fPIC -Werror=unguarded-availability-new -Ilibdatrie -c libdatrie/datrie/trie.c -o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/trie.o
/var/folders/t_/mmhnh941511_hp2lwh383bp00000gn/T/tmpc74tdl4c/tools/llvm/bin/llvm-ar rcs build/temp.macosx-11.0-arm64-cpython-313/libdatrie.a build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/alpha-map.o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/darray.o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/dstring.o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/fileutils.o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/tail.o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/trie-string.o build/temp.macosx-11.0-arm64-cpython-313/libdatrie/datrie/trie.o

[stderr]
/Users/shsingh/.cache/uv/builds-v0/.tmpq5HjgW/lib/python3.13/site-packages/setuptools/_distutils/dist.py:270: UserWarning: Unknown distribution option: 'tests_require'
  warnings.warn(msg)
libdatrie/datrie/alpha-map.c:500:21: warning: comparison of integers of different signs: 'TrieIndex' (aka 'int') and 'AlphaChar' (aka 'unsigned int') [-Wsign-compare]
    if (alpha_begin <= ac && ac <= alpha_map->alpha_end)
        ~~~~~~~~~~~ ^  ~~
1 warning generated.
libdatrie/datrie/darray.c:239:22: warning: comparison of integers of different signs: 'TrieIndex' (aka 'int') and 'unsigned long' [-Wsign-compare]
    if (d->num_cells > SIZE_MAX / sizeof (DACell))
        ~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
libdatrie/datrie/fileutils.c:103:52: warning: comparison of integers of different signs: 'unsigned long' and 'int' [-Wsign-compare]
    return (fread (buff, sizeof (char), len, file) == len);
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~
libdatrie/datrie/fileutils.c:109:53: warning: comparison of integers of different signs: 'unsigned long' and 'int' [-Wsign-compare]
    return (fwrite (buff, sizeof (char), len, file) == len);
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^  ~~~
2 warnings generated.
libdatrie/datrie/tail.c:144:22: warning: comparison of integers of different signs: 'TrieIndex' (aka 'int') and 'unsigned long' [-Wsign-compare]
    if (t->num_tails > SIZE_MAX / sizeof (TailBlock))
        ~~~~~~~~~~~~ ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
error: command '/var/folders/t_/mmhnh941511_hp2lwh383bp00000gn/T/tmpc74tdl4c/tools/llvm/bin/llvm-ar' failed: No such file or directory
```

So I had to do this on OSX 14.7.1 (23H222)

```
export AR=/usr/bin/ar
uv sync
```

shntnu commented:
> There's some issue with some numbers in the output of harmony-pca. seems some of the PCs are flipped `- (x)`,thus the `pd.assert_equal` statement fails.

Per the 🤖 , we can later change this

```py
features = pd.DataFrame(PCA(variance).fit_transform(features))
```

to

```py
# Fit PCA
pca = PCA(variance)
transformed = pca.fit_transform(features)

# Ensure consistent signs by making the largest absolute component positive
for i in range(transformed.shape[1]):
    if np.abs(transformed[np.argmax(np.abs(transformed[:, i])), i]) < 0:
        transformed[:, i] *= -1
        
features = pd.DataFrame(transformed)
```

But let's change the test for now

shntnu commented:
FYI I tried sign flipping in the test and it didn’t seem to be sufficient.

On Thu, 20 Feb 2025 at 3:25 AM John Arevalo ***@***.***>
wrote:

> Thanks for setting all this. I tried to run this in my pc, but mamba
> failed to reproduce the environment. Instead, I moved it to uv as that's
> what we are now using inside the lab. I also updated snakemake to the most
> recent stable version.
>
> There's some issue with some numbers in the output of harmony-pca. seems
> some of the PCs are flipped - (x),thus the pd.assert_equal statement
> fails.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/jump-profiling-recipe/pull/28#issuecomment-2669848814>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJHQPEY6OGTY2ZSNIJ3GZD2QT4WHAVCNFSM6AAAAABXITRRCWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRZHA2DQOBRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: johnarevalo]*johnarevalo* left a comment
> (broadinstitute/jump-profiling-recipe#28)
> <https://github.com/broadinstitute/jump-profiling-recipe/pull/28#issuecomment-2669848814>
>
> Thanks for setting all this. I tried to run this in my pc, but mamba
> failed to reproduce the environment. Instead, I moved it to uv as that's
> what we are now using inside the lab. I also updated snakemake to the most
> recent stable version.
>
> There's some issue with some numbers in the output of harmony-pca. seems
> some of the PCs are flipped - (x),thus the pd.assert_equal statement
> fails.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/broadinstitute/jump-profiling-recipe/pull/28#issuecomment-2669848814>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJHQPEY6OGTY2ZSNIJ3GZD2QT4WHAVCNFSM6AAAAABXITRRCWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMNRZHA2DQOBRGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

johnarevalo commented:
I ugly-patched by taking the `abs()` on both dataframes. This is because not all the columns are sign-flipped in the same dataframe. I'll create a separate issue to follow-up on this. For now I will merge it.

Thanks!

shntnu commented:
Great!
